README.md
# Factorio-Zooming-Out-Screenshots

Takes a series of screenshots steadily zooming out from the player's location.
![Example Screenshot Montage made to Video](https://media.giphy.com/media/kWFmXsGmysEnTUkCSm/giphy.gif)

Features
=========

- Configurable start and target end zoom levels.
- Configurable number of screenshots to take between zoom levels.
- Zooms out at an auto-calculated rate of decay (exponential). Designed for creating zooming out videos or montages.
- Standard screenshot options are exposed in the GUI: resolution, show the GUI, show entity info (alt-mode), anti-aliasing.
- Quick test screenshot option to get single images at the specified zoom level quickly.

Notes
=========

- Minimum Factorio final zoom value is 0.04 since Factorio 0.18, enforced by the mod.
- Screenshots are saved to the player's script-output folder in the format "ZoomingScreenshot_NUMBER" starting at number 1 for the Start Zoom level. Output files are all in the .png file type to ensure quality.
- The Test End Zoom screenshot is just an indication of the final zoom level. There is a rounding error that will cause the last screenshots to be slightly zoomed out more than the specified target, but they will match the test screenshot.
- If doing less than 50~ screenshots the zoom algorithm may not zoom out quite to the End Zoom setting, but will be very close.

Usage Warning
=========

When taking screenshots the game will pause for a while (seconds to minutes) based on the number of screenshot, resolution and zoom. This is standard Factorio behaviour and so it is advised to use this mod only on local games to avoid multiplayer disconnects. In extreme cases, the game may run out of memory and crash or may hang for so long that windows kills it (varies between pcs)

There are some oddities with taking screenshots and certain setting combinations within Factorio. All of these apply equally to this mod.  Fir example; enabling GUI and using a resolution different to your display resolution will create a mis aligned image.


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 20.1.5
Date: 2020-01-26
  Features:
    - updated to Factorio 1.1 as-is
  Changes:
  Bugfixes:
    - Correct default resolution y to 768.
  Info:
---------------------------------------------------------------------------------------------------
Version: 18.1.4
Date: 2020-01-26
  Features:
    - updated to Factorio 0.18 as-is
  Changes:
    - max zoom out locked at 0.04 as this is a reliably working min zoom Facotrio 0.18 supports. done in a clunky way for now.
  Bugfixes:
  Info:
---------------------------------------------------------------------------------------------------
Version: 17.1.3
Date: 2019-03-11
  Features:
    - Updated to support 0.17
    - Added support for the Shortcut bar to open the mods GUI.
  Changes:
    - End Zoom changed to be Target End Zoom. The Target End Zoom Test Screenshot button now uses the Take All Screenshot maths to calculate the end zoom. So the test will match the real screenshots.
    - Removed the Toggle mod GUI button from top left.
  Info:
    - changelog changed to new format
    - minor refactoring of code
---------------------------------------------------------------------------------------------------
Version: 0.1.2
Date: 2019-02-12
  Bugfixes:
    - Fixed bug with laoding the mod on to an existing map that could cause an error on use of the mod show gui button
---------------------------------------------------------------------------------------------------
Version: 0.1.1
Date: 2019-01-29
  Changes:
    - Screenshots only taken by the player who uses the GUI.
  Bugfixes:
    - Fixed that the Test Screenshot buttons did nothing when clicked
    - Fixed some missing tooltips
    - When Taking All Screenshots it waits a tick for the gui to disappear before the screenshots are taken
  Info:
    - Added example zoom out video link to readme
---------------------------------------------------------------------------------------------------
Version: 0.1.0
Date: 2019-01-28
  Features:
    - Initial Release

data.lua
local Constants = require("constants")

data:extend(
    {
        {
            type = "shortcut",
            name = "zooming-screenshot",
            action = "lua",
            icon = {
                filename = Constants.AssetModName .. "/graphics/gui/camera-button.png",
                width = 36,
                height = 36
            },
            small_icon = {
                filename = Constants.AssetModName .. "/graphics/gui/camera-button.png",
                width = 36,
                height = 36
            },
            disabled_small_icon = {
                filename = Constants.AssetModName .. "/graphics/gui/camera-button-disabled.png",
                width = 36,
                height = 36
            }
        },
        {
            type = "sprite",
            name = "zooming-screenshit-mod-icon",
            filename = Constants.AssetModName .. "/graphics/gui/camera-button.png",
            width = 36,
            height = 36
        }
    }
)

local defaultStyle = data.raw["gui-style"]["default"]

defaultStyle.zs_padded_horizontal_flow = {
    type = "horizontal_flow_style",
    left_padding = 4,
    top_padding = 4
}

defaultStyle.zs_mod_button_sprite = {
    type = "button_style",
    width = 36,
    height = 36,
    scalable = true
}

--same size as a button
defaultStyle.zs_button_sprite = {
    type = "button_style",
    width = 42,
    height = 42,
    scalable = true
}

defaultStyle.zs_gui_frame = {
    type = "frame_style",
    left_padding = 4,
    top_padding = 4
}

defaultStyle.zs_padded_table = {
    type = "table_style",
    top_padding = 5,
    bottom_padding = 5,
    left_padding = 5,
    right_padding = 5
}

defaultStyle.zs_padded_table_cell = {
    type = "label_style",
    top_padding = 5,
    bottom_padding = 5,
    left_padding = 5,
    right_padding = 5
}

defaultStyle.zs_small_button = {
    type = "button_style",
    padding = 2,
    font = "default"
}


control.lua
local Gui = require("gui")

local function MakeGlobals()
    if global.MOD == nil then
        global.MOD = {}
    end
    if global.MOD.guiScreenshotSettings == nil then
        global.MOD.guiScreenshotSettings = {}
    end
end

local function PlayerJoinedEvent(eventData)
    Gui.GenerateDefaultSettings(eventData.player_index)
end

local function OnInit()
    MakeGlobals()
    Gui.CloseGuiForAll()
end

local function OnConfigurationChanged()
    MakeGlobals()
    Gui.GenerateDefaultSettingsForAll()
    Gui.CloseGuiForAll()
end

script.on_init(OnInit)
script.on_configuration_changed(OnConfigurationChanged)
script.on_event(defines.events.on_player_joined_game, PlayerJoinedEvent)
script.on_event(defines.events.on_gui_click, Gui.GuiClickedEvent)
script.on_event(defines.events.on_gui_text_changed, Gui.GuiTextChangedEvent)
script.on_event(defines.events.on_gui_selection_state_changed, Gui.GuiSelectionStateChangedEvent)
script.on_event(defines.events.on_lua_shortcut, Gui.ShortcutClicked)


screenshot.lua
local Screenshot = {}

function Screenshot.CalculateZoomDetails(player)
    local screenshotSettings = global.MOD.guiScreenshotSettings[player.index]
    local startZoom = screenshotSettings["zooming-screenshot-start-zoom"]
    local targetEndZoom = screenshotSettings["zooming-screenshot-end-zoom"]
    if tonumber(targetEndZoom) < 0.04 then
        targetEndZoom = 0.04
    end
    local screenshotCount = screenshotSettings["zooming-screenshot-count"]
    local dividedValue = targetEndZoom / startZoom
    local naturalLog = math.log(dividedValue)
    local decay = naturalLog / (screenshotCount - 1)
    local zoomDevider = 1 + decay
    return {zoomDevider = zoomDevider, screenshotCount = screenshotCount, startZoom = startZoom, targetEndZoom = targetEndZoom}
end

function Screenshot.TakeScheduledScreenshots(player)
    script.on_nth_tick(game.tick, nil)
    if not player.valid then
        return
    end
    local zoomDetails = Screenshot.CalculateZoomDetails(player)
    local zoom = zoomDetails.startZoom
    for shotCount = 1, zoomDetails.screenshotCount do
        Screenshot.TakeScreenshot(player, zoom, "ZoomingScreenshot_" .. shotCount)
        zoom = zoom * zoomDetails.zoomDevider
    end
end

function Screenshot.TakeAllScreenShots(player)
    local triggerTick = game.tick + 1
    script.on_nth_tick(
        triggerTick,
        function()
            Screenshot.TakeScheduledScreenshots(player)
        end
    )
end

function Screenshot.TakeStartZoomTestScreenshot(player)
    local zoomDetails = Screenshot.CalculateZoomDetails(player)
    Screenshot.TakeScreenshot(player, zoomDetails.startZoom, "Start_Zoom_" .. zoomDetails.startZoom)
end

function Screenshot.TakeEndZoomTestScreenshot(player)
    local zoomDetails = Screenshot.CalculateZoomDetails(player)
    local zoom = zoomDetails.startZoom
    for i = 1, zoomDetails.screenshotCount - 1 do
        zoom = zoom * zoomDetails.zoomDevider
    end
    Screenshot.TakeScreenshot(player, zoom, "End_Zoom_" .. zoomDetails.targetEndZoom)
end

function Screenshot.TakeScreenshot(player, zoom, name)
    local screenshotSettings = global.MOD.guiScreenshotSettings[player.index]
    local screenshotPosition = {x = player.position.x, y = player.position.y - 0.6}
    game.take_screenshot {
        by_player = player,
        position = screenshotPosition,
        resolution = {
            x = screenshotSettings["zooming-screenshot-resolution-x"],
            y = screenshotSettings["zooming-screenshot-resolution-y"]
        },
        zoom = zoom,
        show_gui = Screenshot.BooleanDropdownIndexToBool(screenshotSettings["zooming-screenshot-show-gui"]),
        show_entity_info = Screenshot.BooleanDropdownIndexToBool(screenshotSettings["zooming-screenshot-show-entity-info"]),
        anti_alias = Screenshot.BooleanDropdownIndexToBool(screenshotSettings["zooming-screenshot-anti-alias"]),
        path = name .. ".png"
    }
end

function Screenshot.BooleanDropdownIndexToBool(value)
    if value == 1 then
        return true
    elseif value == 2 then
        return false
    else
        return nil
    end
end

return Screenshot


gui.lua
local Screenshot = require("screenshot")
local Gui = {}

function Gui.GenerateDefaultSettings(playerID)
    if global.MOD.guiScreenshotSettings[playerID] == nil then
        global.MOD.guiScreenshotSettings[playerID] = {}
    end
    local guiScreenshotSettings = global.MOD.guiScreenshotSettings[playerID]
    if guiScreenshotSettings["zooming-screenshot-start-zoom"] == nil then
        guiScreenshotSettings["zooming-screenshot-start-zoom"] = 20
    end
    if guiScreenshotSettings["zooming-screenshot-end-zoom"] == nil then
        guiScreenshotSettings["zooming-screenshot-end-zoom"] = 0.125
    end
    if guiScreenshotSettings["zooming-screenshot-count"] == nil then
        guiScreenshotSettings["zooming-screenshot-count"] = 100
    end
    if guiScreenshotSettings["zooming-screenshot-resolution-x"] == nil then
        guiScreenshotSettings["zooming-screenshot-resolution-x"] = 1024
    end
    if guiScreenshotSettings["zooming-screenshot-resolution-y"] == nil then
        guiScreenshotSettings["zooming-screenshot-resolution-y"] = 768
    end
    if guiScreenshotSettings["zooming-screenshot-show-gui"] == nil then
        guiScreenshotSettings["zooming-screenshot-show-gui"] = 2
    end
    if guiScreenshotSettings["zooming-screenshot-show-entity-info"] == nil then
        guiScreenshotSettings["zooming-screenshot-show-entity-info"] = 2
    end
    if guiScreenshotSettings["zooming-screenshot-anti-alias"] == nil then
        guiScreenshotSettings["zooming-screenshot-anti-alias"] = 1
    end
end

function Gui.GenerateDefaultSettingsForAll()
    for _, player in pairs(game.players) do
        Gui.GenerateDefaultSettings(player.index)
    end
end

function Gui.GuiClickedEvent(eventData)
    local player = game.get_player(eventData.player_index)
    local clickedElement = eventData.element
    if clickedElement.name == "close-zooming-screenshot-gui-button" then
        Gui.CloseGui(player)
    elseif clickedElement.name == "zooming-screenshot-take-all-button" then
        Gui.TakeAllScreenShots(player)
    elseif clickedElement.name == "zooming-screenshot-start-zoom-test-screenshot" then
        Screenshot.TakeStartZoomTestScreenshot(player)
    elseif clickedElement.name == "zooming-screenshot-end-zoom-test-screenshot" then
        Screenshot.TakeEndZoomTestScreenshot(player)
    end
end

function Gui.GuiTextChangedEvent(eventData)
    local editedElement = eventData.element
    local screenshotSettings = global.MOD.guiScreenshotSettings[eventData.player_index]
    if screenshotSettings[editedElement.name] ~= nil then
        screenshotSettings[editedElement.name] = editedElement.text
    end
    local endZoomNumber = tonumber(screenshotSettings["zooming-screenshot-end-zoom"])
    if endZoomNumber ~= nil and endZoomNumber < 0.04 then
        local player = game.get_player(eventData.player_index)
        local guiElement = player.gui.left["zooming-screenshot-gui-flow"]["zooming-screenshot-gui-frame"]["zooming-screenshot-options-table"]["zooming-screenshot-end-zoom"]
        guiElement.text = "0.04"
        screenshotSettings["zooming-screenshot-end-zoom"] = "0.04"
    end
end

function Gui.GuiSelectionStateChangedEvent(eventData)
    local editedElement = eventData.element
    local screenshotSettings = global.MOD.guiScreenshotSettings[eventData.player_index]
    if screenshotSettings[editedElement.name] ~= nil then
        screenshotSettings[editedElement.name] = editedElement.selected_index
    end
end

function Gui.ShortcutClicked(eventData)
    local shortcutName = eventData.prototype_name
    if shortcutName ~= "zooming-screenshot" then
        return
    end
    local player = game.get_player(eventData.player_index)
    Gui.ToggleGui(player)
end

function Gui.CloseGuiForAll()
    for _, player in pairs(game.players) do
        Gui.CloseGui(player)
    end
end

function Gui.CloseGui(player)
    local modGui = player.gui.left["zooming-screenshot-gui-flow"]
    if modGui ~= nil then
        modGui.destroy()
    end
end

function Gui.ToggleGui(player)
    if player.gui.left["zooming-screenshot-gui-flow"] == nil then
        Gui.OpenGui(player)
    else
        Gui.CloseGui(player)
    end
end

function Gui.OpenGui(player)
    local screenshotSettings = global.MOD.guiScreenshotSettings[player.index]
    Gui.CloseGui(player)
    local guiFlow = player.gui.left.add {type = "flow", name = "zooming-screenshot-gui-flow", style = "zs_padded_horizontal_flow", direction = "horizontal"}
    local guiFrame = guiFlow.add {type = "frame", name = "zooming-screenshot-gui-frame", direction = "vertical", style = "zs_gui_frame"}

    local optionsTable = guiFrame.add {type = "table", name = "zooming-screenshot-options-table", column_count = 3}

    optionsTable.add {type = "label", name = "zooming-screenshot-start-zoom-label", caption = {"gui-caption.start-zoom"}, tooltip = {"gui-tooltip.start-zoom"}}
    optionsTable.add {type = "textfield", name = "zooming-screenshot-start-zoom", text = screenshotSettings["zooming-screenshot-start-zoom"]}
    optionsTable.add {type = "button", name = "zooming-screenshot-start-zoom-test-screenshot", caption = {"gui-caption.start-zoom-test-screenshot"}, tooltip = {"gui-tooltip.start-zoom-test-screenshot"}, style = "zs_small_button"}

    optionsTable.add {type = "label", name = "zooming-screenshot-end-zoom-label", caption = {"gui-caption.end-zoom"}, tooltip = {"gui-tooltip.end-zoom"}}
    optionsTable.add {type = "textfield", name = "zooming-screenshot-end-zoom", text = screenshotSettings["zooming-screenshot-end-zoom"]}
    optionsTable.add {type = "button", name = "zooming-screenshot-end-zoom-test-screenshot", caption = {"gui-caption.end-zoom-test-screenshot"}, tooltip = {"gui-tooltip.end-zoom-test-screenshot"}, style = "zs_small_button"}

    optionsTable.add {type = "label", name = "zooming-screenshot-count-label", caption = {"gui-caption.screenshot-count"}, tooltip = {"gui-tooltip.screenshot-count"}}
    optionsTable.add {type = "textfield", name = "zooming-screenshot-count", text = screenshotSettings["zooming-screenshot-count"]}
    optionsTable.add {type = "label", name = "zooming-screenshot-count-padder"}

    local spacer = optionsTable.add {type = "label", name = "zooming-screenshot-spacer1"}
    spacer.style.height = 10
    optionsTable.add {type = "label", name = "zooming-screenshot-spacer2"}
    optionsTable.add {type = "label", name = "zooming-screenshot-spacer3"}

    optionsTable.add {type = "label", name = "zooming-screenshot-resolution-label", caption = {"gui-caption.resolution"}, tooltip = {"gui-tooltip.resolution"}}
    local resolutionFlow = optionsTable.add {type = "flow", name = "zooming-screenshot-resolution-flow", direction = "horizontal"}
    resolutionFlow.style.width = 150
    local resolutionXInput = resolutionFlow.add {type = "textfield", name = "zooming-screenshot-resolution-x", text = screenshotSettings["zooming-screenshot-resolution-x"], tooltip = {"gui-tooltip.resolution-x"}}
    resolutionXInput.style.width = 70
    local resolutionRightFlow = resolutionFlow.add {type = "flow", name = "zooming-screenshot-resolution-flow-right", direction = "horizontal"}
    resolutionRightFlow.style.horizontal_align = "right"
    resolutionRightFlow.style.horizontally_stretchable = true
    local resolutionYInput = resolutionRightFlow.add {type = "textfield", name = "zooming-screenshot-resolution-y", text = screenshotSettings["zooming-screenshot-resolution-y"], tooltip = {"gui-tooltip.resolution-y"}}
    resolutionYInput.style.width = 70
    optionsTable.add {type = "label", name = "zooming-screenshot-resolution-padder"}

    optionsTable.add {type = "label", name = "zooming-screenshot-show-gui-label", caption = {"gui-caption.show-gui"}, tooltip = {"gui-tooltip.show-gui"}}
    optionsTable.add {type = "drop-down", name = "zooming-screenshot-show-gui", items = {"Yes", "No"}, selected_index = screenshotSettings["zooming-screenshot-show-gui"]}
    optionsTable.add {type = "label", name = "zooming-screenshot-show-gui-padder"}

    optionsTable.add {type = "label", name = "zooming-screenshot-show-entity-info-label", caption = {"gui-caption.show-entity-info"}, tooltip = {"gui-tooltip.show-entity-info"}}
    optionsTable.add {type = "drop-down", name = "zooming-screenshot-show-entity-info", items = {"Yes", "No"}, selected_index = screenshotSettings["zooming-screenshot-show-entity-info"]}
    optionsTable.add {type = "label", name = "zooming-screenshot-show-entity-info-padder"}

    optionsTable.add {type = "label", name = "zooming-screenshot-anti-alias-label", caption = {"gui-caption.anti-alias"}, tooltip = {"gui-tooltip.anti-alias"}}
    optionsTable.add {type = "drop-down", name = "zooming-screenshot-anti-alias", items = {"Yes", "No"}, selected_index = screenshotSettings["zooming-screenshot-anti-alias"]}
    optionsTable.add {type = "label", name = "zooming-screenshot-anti-alias-padder"}

    local bottomSection = guiFrame.add {type = "flow", name = "zooming-screenshot-bottom-section", direction = "horizontal"}
    bottomSection.style.top_padding = 4
    bottomSection.add {type = "button", name = "close-zooming-screenshot-gui-button", caption = {"gui-caption.close-gui-button"}}
    local bottomSectionRightFlow = bottomSection.add {type = "flow", name = "zooming-screenshot-bottom-section-right-flow"}
    bottomSectionRightFlow.style.horizontally_stretchable = true
    bottomSectionRightFlow.style.horizontal_align = "right"
    bottomSectionRightFlow.add {type = "button", name = "zooming-screenshot-take-all-button", caption = {"gui-caption.take-all-screenshots-button"}}
end

function Gui.TakeAllScreenShots(player)
    Gui.CloseGui(player)
    Screenshot.TakeAllScreenShots(player)
end

return Gui
